docs/config.md
# 配置项

## 登录设置

### type

- 类型：`'login' | 'token' | 'naifu' | 'sd-webui' | 'stable-horde'`
- 默认值：`'token'`

登录方式。`login` 表示使用账号密码登录，`token` 表示使用授权令牌登录。`naifu`、`sd-webui` 和 `stable-horde` 对应着其他类型的后端。

### email

- 类型：`string`
- 当 `type` 为 `login` 时必填

你的账号邮箱。

### password

- 类型：`string`
- 当 `type` 为 `login` 时必填

你的账号密码。

### token

- 类型：`string`
- 当 `type` 为 `token` 时必填

授权令牌。获取方式如下：

1. 在网页中登录你的 NovelAI 账号
2. 打开控制台 (F12)，并切换到控制台 (Console) 标签页
3. 输入下面的代码并按下回车运行

```js
console.log(JSON.parse(localStorage.session).auth_token)
```

4. 输出的字符串就是你的授权令牌

### endpoint

- 类型：`string`
- 默认值：`'https://api.novelai.net'`
- 当 `type` 为 `naifu` 或 `sd-webui` 时必填

API 服务器地址。如果你搭建了私服，可以将此项设置为你的服务器地址。

### headers

- 类型：`Dict<string>`
- 默认值：官服的 `Referer` 和 `User-Agent`

要附加的额外请求头。如果你的 `endpoint` 是第三方服务器，你可能需要设置正确的请求头，否则请求可能会被拒绝。

### trustedWorkers

- 类型: `boolean`
- 默认值: `false`
- 当 `type` 为 `stable-horde` 时可选

是否只请求可信任工作节点。

### pollInterval

- 类型: `number`
- 默认值: `1000`
- 当 `type` 为 `stable-horde` 时可选

轮询进度间隔时长。单位为毫秒。

## 参数设置

### model

- 类型：`'safe' | 'nai' | 'furry'`
- 默认值：`'nai'`

默认的生成模型。

### sampler

- 类型：`'k_euler_ancestral' | 'k_euler' | 'k_lms' | 'plms' | 'ddim'`
- 默认值：`'k_euler_ancestral'`

默认的采样器。

### scale

- 类型：`number`
- 默认值：`11`

默认对输入的服从度。

### textSteps

- 类型：`number`
- 默认值：`28`

文本生图时默认的迭代步数。

### imageSteps

- 类型：`number`
- 默认值：`50`

以图生图时默认的迭代步数。

### maxSteps

- 类型：`number`
- 默认值：`64`

允许的最大迭代步数。

### strength

- 类型：`number`
- 默认值：`0.7`
- 取值范围：`(0, 1]`

默认的重绘强度。

### resolution

- 类型：`'portrait' | 'square' | 'landscape' | { width: number, height: number }`
- 默认值：`'portrait'`

默认生成的图片尺寸。

### maxResolution

- 类型：`number`
- 默认值：`1024`

允许生成的宽高最大值。

## 输入设置

### basePrompt

- 类型: `string`
- 默认值: `'masterpiece, best quality'`

所有请求的附加标签。默认值相当于网页版的「Add Quality Tags」功能。

### negativePrompt

- 类型: `string`
- 默认值:
  ```text
  nsfw, lowres, bad anatomy, bad hands, text, error, missing fingers,
  extra digit, fewer digits, cropped, worst quality, low quality,
  normal quality, jpeg artifacts, signature, watermark, username, blurry
  ```

所有请求附加的负面标签。默认值相当于网页版的「Low Quality + Bad Anatomy」排除。

### forbidden

- 类型：`string`
- 默认值：`''`

违禁词列表。请求中的违禁词将会被自动删除。

违禁词语法与关键词类似，使用逗号隔开英文单词。由于它只用于过滤输入，因此不接受影响因子和要素混合等高级语法。默认情况下，每个违禁词均采用模糊匹配，即只要输入的某个关键词中包含任何一个违禁词作为子串，就会被自动删除。如果要使用精确匹配，可以在词尾加上 `!`。例如 `sex!` 仅会过滤 `sex toys` 而不过滤 `sexy girl`。

默认情况下本插件不设违禁词。对于想要禁用 nsfw 内容的用户，下面的违禁词表可供参考：

```text
guro, nipple, anal, anus, masturbation, sex!, rape, fuck,
dick, testis, nude, nake, cum, nudity, virgina, penis, nsfw,
topless, ass, bdsm, footjob, handjob, fellatio, deepthroat,
cum, ejaculation, bukkake, orgasm, pussy, bloody
```

### placement

- 类型：`'before' | 'after'`
- 默认值：`'after'`

默认附加标签相对用户输入的摆放位置。

设置为 `before` 意味着默认标签拥有更高的优先级 (如果希望与 NovelAI 官网行为保持一致，推荐这个选项)，而设置为 `after` 将允许用户更高的创作自由度。

在 `before` 模式下，用户仍然可以通过 `-O, --override` 选项手动忽略默认标签；而在 `after` 模式下，用户仍然可以通过将基础标签写在前面的方式手动提高基础标签的优先级。

### translator

- 类型：`boolean`
- 默认值：`true`

是否启用自动翻译。安装任意 [翻译插件](https://translator.koishi.chat) 后即可自动将用户输入转换为英文。

### latinOnly

- 类型：`boolean`
- 默认值：`false`

是否只接受英文单词。启用后将对于非拉丁字母的输入进行错误提示。

::: tip
[自动翻译](#translator) 会转化用户输入，因此不建议与此选项同时启用。
:::

### maxWords

- 类型：`number`
- 默认值：`0`

用户输入的最大词数 (设置为 `0` 时无限制)。下面是一些细节设定：

- 这个配置项限制的是词数而不是标签数或字符数 (`girl, pleated skirt` 的词数为 3)
- 正向标签和反向标签分别计数，每种均不能超出此限制，但总词数可以达到限制的两倍
- 如果开启了 [自动翻译](#translator)，此限制将作用于翻译后的结果
- 默认附加的正向和反向标签均不计入此限制

## 高级设置

### output

- 类型：`'minimal' | 'default' | 'verbose'`
- 默认值：`'default'`

输出方式。`minimal` 表示只发送图片，`default` 表示发送图片和关键信息，`verbose` 表示发送全部信息。

### allowAnlas

- 类型：`boolean | number`
- 默认值：`true`

是否启用高级功能。高级功能包括：

- 以图生图相关功能
- `-r, --resolution` 选项中，手动指定具体宽高
- `-O, --override` 选项
- `-t, --steps` 选项
- `-n, --noise` 选项
- `-N, --strength` 选项

当设置为数字时，表示使用上述功能所需的最低权限等级。

### requestTimeout

- 类型：`number`
- 默认值：`30000`

当请求超过这个时间时会中止并提示超时。

### maxRetryCount

- 类型：`number`
- 默认值：`3`

连接失败时最大的重试次数。

<!-- ### recallTimeout

- 类型：`number`
- 默认值：`0`

图片发送后自动撤回的时间 (设置为 `0` 禁用此功能)。 -->

### maxIterations

- 类型：`number`
- 默认值：`1`

允许的最大绘制次数。参见 [批量生成](./usage.md#批量生成)。

### maxConcurrency

- 类型：`number`
- 默认值：`0`

单个频道下的最大并发数量 (设置为 `0` 以禁用此功能)。


docs/advanced/config.md
# 配置文件

每个 Koishi 应用都有一个配置文件，它管理了应用及其插件的全部配置。配置文件的格式为 [YAML](https://en.wikipedia.org/wiki/YAML)，它是一种易于阅读和编辑的文本格式，你可以用任何文本编辑器打开。

::: tip
如果你不了解 YAML 的语法，请不要随意修改配置文件，否则将可能导致 Koishi 应用无法运行。你可以在[这篇教程](https://www.runoob.com/w3cnote/yaml-intro.html)中学习 YAML 的语法。
:::

## 根目录位置

配置文件所在的目录叫**根目录**。根据你的安装方式，根目录的位置可能不同：

- zip: 解压目录下 `data/instances/default`
- msi: `C:/Users/你的用户名/AppData/Roaming/Koishi/Desktop/data/instances/default`
- pkg: `~/Library/Application Support/Koishi/Desktop/data/instances/default`

配置文件是根目录下名为 `koishi.yml` 的文件。当你遇到问题时，开发者可能会要求你提供配置文件的内容。此时去上面的地方找就好了。

## 理解配置文件

尝试打开配置文件，你会发现它的内容大致如下：

```yaml
# 全局设置
host: localhost
port: 5140

# 插件列表
plugins:
  # group 表示这是一个插件组
  group:basic:
    help:
    rate-limit:
    locales:
    commands:

  # 这是另一个插件组
  group:adapter:
    # 波浪线前缀表示一个不启用的插件
    ~adapter-onebot:
    ~adapter-discord:
    ~adapter-telegram:
    gocqhttp:

  # 你刚刚安装的 NovelAI 插件
  novelai:
```

你会发现，配置文件的结构与「插件配置」页面基本是一致的。当你启动 Koishi 应用时，Koishi 会读取上述配置文件并加载所需的插件；而当你在「插件配置」页面中修改了某些配置，Koishi 也会自动将这些改动写入配置文件。

绝大多数的功能都可以通过「插件配置」页面来完成，但目前尚有一些功能没有做好相应的交互界面，这时你就需要手动修改配置文件了。你需要做的有以下几步：

1. 关闭当前 Koishi 应用
2. 打开配置文件进行编辑
3. 保存配置文件后再次启动 Koishi 应用


src/config.ts
import { Computed, Dict, Schema, Session, Time } from 'koishi'
import { Size } from './utils'

const options: Computed.Options = {
  userFields: ['authority'],
}

export const modelMap = {
  safe: 'safe-diffusion',
  nai: 'nai-diffusion',
  furry: 'nai-diffusion-furry',
} as const

export const orientMap = {
  landscape: { height: 512, width: 768 },
  portrait: { height: 768, width: 512 },
  square: { height: 640, width: 640 },
} as const

export const hordeModels = require('../data/horde-models.json') as string[]

const ucPreset = [
  'nsfw, lowres, bad anatomy, bad hands, text, error, missing fingers',
  'extra digit, fewer digits, cropped, worst quality, low quality',
  'normal quality, jpeg artifacts, signature, watermark, username, blurry',
].join(', ')

type Model = keyof typeof modelMap
type Orient = keyof typeof orientMap

export const models = Object.keys(modelMap) as Model[]
export const orients = Object.keys(orientMap) as Orient[]

export namespace sampler {
  export const nai = {
    'k_euler_a': 'Euler ancestral',
    'k_euler': 'Euler',
    'k_lms': 'LMS',
    'ddim': 'DDIM',
    'plms': 'PLMS',
  }

  // samplers in stable-diffusion-webui
  // https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/master/modules/sd_samplers_compvis.py#L12
  // https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/master/modules/sd_samplers_kdiffusion.py#L12
  export const sd = {
    'k_euler_a': 'Euler a',
    'k_euler': 'Euler',
    'k_lms': 'LMS',
    'k_heun': 'Heun',
    'k_dpm_2': 'DPM2',
    'k_dpm_2_a': 'DPM2 a',
    'k_dpmpp_2s_a': 'DPM++ 2S a',
    'k_dpmpp_2m': 'DPM++ 2M',
    'k_dpmpp_sde': 'DPM++ SDE',
    'k_dpmpp_2m_sde': 'DPM++ 2M SDE',
    'k_dpm_fast': 'DPM fast',
    'k_dpm_ad': 'DPM adaptive',
    'k_lms_ka': 'LMS Karras',
    'k_dpm_2_ka': 'DPM2 Karras',
    'k_dpm_2_a_ka': 'DPM2 a Karras',
    'k_dpmpp_2s_a_ka': 'DPM++ 2S a Karras',
    'k_dpmpp_2m_ka': 'DPM++ 2M Karras',
    'k_dpmpp_sde_ka': 'DPM++ SDE Karras',
    'k_dpmpp_2m_sde_ka': 'DPM++ 2M SDE Karras',
    'ddim': 'DDIM',
    'plms': 'PLMS',
    'unipc': 'UniPC',
  }

  export const horde = {
    k_lms: 'LMS',
    k_heun: 'Heun',
    k_euler: 'Euler',
    k_euler_a: 'Euler a',
    k_dpm_2: 'DPM2',
    k_dpm_2_a: 'DPM2 a',
    k_dpm_fast: 'DPM fast',
    k_dpm_adaptive: 'DPM adaptive',
    k_dpmpp_2m: 'DPM++ 2M',
    k_dpmpp_2s_a: 'DPM++ 2S a',
    k_lms_ka: 'LMS Karras',
    k_heun_ka: 'Heun Karras',
    k_euler_ka: 'Euler Karras',
    k_euler_a_ka: 'Euler a Karras',
    k_dpm_2_ka: 'DPM2 Karras',
    k_dpm_2_a_ka: 'DPM2 a Karras',
    k_dpm_fast_ka: 'DPM fast Karras',
    k_dpm_adaptive_ka: 'DPM adaptive Karras',
    k_dpmpp_2m_ka: 'DPM++ 2M Karras',
    k_dpmpp_2s_a_ka: 'DPM++ 2S a Karras',
  }

  export function createSchema(map: Dict<string>) {
    return Schema.union(Object.entries(map).map(([key, value]) => {
      return Schema.const(key).description(value)
    })).loose().description('默认的采样器。').default('k_euler_a')
  }

  export function sd2nai(sampler: string): string {
    if (sampler === 'k_euler_a') return 'k_euler_ancestral'
    if (sampler in nai) return sampler
    return 'k_euler_ancestral'
  }
}

export const upscalers = [
  // built-in upscalers
  'None',
  'Lanczos',
  'Nearest',
  // third-party upscalers (might not be available)
  'LDSR',
  'ESRGAN_4x',
  'R-ESRGAN General 4xV3',
  'R-ESRGAN General WDN 4xV3',
  'R-ESRGAN AnimeVideo',
  'R-ESRGAN 4x+',
  'R-ESRGAN 4x+ Anime6B',
  'R-ESRGAN 2x+',
  'ScuNET GAN',
  'ScuNET PSNR',
  'SwinIR 4x',
] as const

export interface Options {
  enhance: boolean
  model: string
  resolution: Size
  sampler: string
  seed: string
  steps: number
  scale: number
  noise: number
  strength: number
}

export interface PromptConfig {
  basePrompt?: Computed<string>
  negativePrompt?: Computed<string>
  defaultPrompt?: Computed<string>
  forbidden?: Computed<string>
  placement?: Computed<'before' | 'after'>
  latinOnly?: Computed<boolean>
  translator?: boolean
  maxWords?: Computed<number>
}

export const PromptConfig: Schema<PromptConfig> = Schema.object({
  basePrompt: Schema.computed(Schema.string().role('textarea'), options).description('默认附加的标签。').default('masterpiece, best quality'),
  negativePrompt: Schema.computed(Schema.string().role('textarea'), options).description('默认附加的反向标签。').default(ucPreset),
  defaultPrompt: Schema.string().role('textarea', options).description('默认提示词，在用户无输入prompt时仍然可以调用。建议在sd-webui中安装dynamic prompt插件，配合使用以达到随机提示词效果。').default(''),
  forbidden: Schema.computed(Schema.string().role('textarea'), options).description('违禁词列表。请求中的违禁词将会被自动删除。').default(''),
  placement: Schema.computed(Schema.union([
    Schema.const('before').description('置于最前'),
    Schema.const('after').description('置于最后'),
  ]), options).description('默认附加标签的位置。').default('after'),
  translator: Schema.boolean().description('是否启用自动翻译。').default(true),
  latinOnly: Schema.computed(Schema.boolean(), options).description('是否只接受英文输入。').default(false),
  maxWords: Schema.computed(Schema.natural(), options).description('允许的最大单词数量。').default(0),
}).description('输入设置')

interface FeatureConfig {
  anlas?: Computed<boolean>
  text?: Computed<boolean>
  image?: Computed<boolean>
  upscale?: Computed<boolean>
}

const naiFeatures = Schema.object({
  anlas: Schema.computed(Schema.boolean(), options).default(true).description('是否允许使用点数。'),
})

const sdFeatures = Schema.object({
  upscale: Schema.computed(Schema.boolean(), options).default(true).description('是否启用图片放大。'),
})

const features = Schema.object({
  text: Schema.computed(Schema.boolean(), options).default(true).description('是否启用文本转图片。'),
  image: Schema.computed(Schema.boolean(), options).default(true).description('是否启用图片转图片。'),
})

interface ParamConfig {
  model?: Model
  sampler?: string
  upscaler?: string
  restoreFaces?: boolean
  hiresFix?: boolean
  scale?: Computed<number>
  textSteps?: Computed<number>
  imageSteps?: Computed<number>
  maxSteps?: Computed<number>
  strength?: Computed<number>
  resolution?: Computed<Orient | Size>
  maxResolution?: Computed<number>
}

export interface Config extends PromptConfig, ParamConfig {
  type: 'token' | 'login' | 'naifu' | 'sd-webui' | 'stable-horde'
  token?: string
  email?: string
  password?: string
  output?: Computed<'minimal' | 'default' | 'verbose'>
  features?: FeatureConfig
  endpoint?: string
  headers?: Dict<string>
  nsfw?: Computed<'disallow' | 'censor' | 'allow'>
  maxIterations?: number
  maxRetryCount?: number
  requestTimeout?: number
  recallTimeout?: number
  maxConcurrency?: number
  pollInterval?: number
  trustedWorkers?: boolean
}

export const Config = Schema.intersect([
  Schema.object({
    type: Schema.union([
      Schema.const('token').description('授权令牌'),
      ...process.env.KOISHI_ENV === 'browser' ? [] : [Schema.const('login').description('账号密码')],
      Schema.const('naifu').description('naifu'),
      Schema.const('sd-webui').description('sd-webui'),
      Schema.const('stable-horde').description('Stable Horde'),
    ]).default('token').description('登录方式。'),
  }).description('登录设置'),

  Schema.union([
    Schema.intersect([
      Schema.union([
        Schema.object({
          type: Schema.const('token'),
          token: Schema.string().description('授权令牌。').role('secret').required(),
        }),
        Schema.object({
          type: Schema.const('login'),
          email: Schema.string().description('账号邮箱。').required(),
          password: Schema.string().description('账号密码。').role('secret').required(),
        }),
      ]),
      Schema.object({
        endpoint: Schema.string().description('API 服务器地址。').default('https://api.novelai.net'),
        headers: Schema.dict(String).role('table').description('要附加的额外请求头。').default({
          'referer': 'https://novelai.net/',
          'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Safari/537.36',
        }),
      }),
    ]),
    Schema.object({
      type: Schema.const('naifu'),
      token: Schema.string().description('授权令牌。').role('secret'),
      endpoint: Schema.string().description('API 服务器地址。').required(),
      headers: Schema.dict(String).role('table').description('要附加的额外请求头。'),
    }),
    Schema.object({
      type: Schema.const('sd-webui'),
      endpoint: Schema.string().description('API 服务器地址。').required(),
      headers: Schema.dict(String).role('table').description('要附加的额外请求头。'),
    }),
    Schema.object({
      type: Schema.const('stable-horde'),
      endpoint: Schema.string().description('API 服务器地址。').default('https://stablehorde.net/'),
      token: Schema.string().description('授权令牌 (API Key)。').role('secret').default('0000000000'),
      nsfw: Schema.union([
        Schema.const('disallow').description('禁止'),
        Schema.const('censor').description('屏蔽'),
        Schema.const('allow').description('允许'),
      ]).description('是否允许 NSFW 内容。').default('allow'),
      trustedWorkers: Schema.boolean().description('是否只请求可信任工作节点。').default(false),
      pollInterval: Schema.number().role('time').description('轮询进度间隔时长。').default(Time.second),
    }),
  ]),

  Schema.object({
    features: Schema.object({}),
  }).description('功能设置'),

  Schema.union([
    Schema.object({
      type: Schema.union(['token', 'login']).hidden(),
      features: Schema.intersect([naiFeatures, features]),
    }),
    Schema.object({
      type: Schema.const('sd-webui'),
      features: Schema.intersect([features, sdFeatures]),
    }),
    Schema.object({
      features: Schema.intersect([features]),
    }),
  ]),

  Schema.object({}).description('参数设置'),

  Schema.union([
    Schema.object({
      type: Schema.const('sd-webui').required(),
      sampler: sampler.createSchema(sampler.sd),
      upscaler: Schema.union(upscalers).description('默认的放大算法。').default('Lanczos'),
      restoreFaces: Schema.boolean().description('是否启用人脸修复。').default(false),
      hiresFix: Schema.boolean().description('是否启用高分辨率修复。').default(false),
    }),
    Schema.object({
      type: Schema.const('stable-horde').required(),
      sampler: sampler.createSchema(sampler.horde),
      model: Schema.union(hordeModels).loose().description('默认的生成模型。'),
    }),
    Schema.object({
      type: Schema.const('naifu').required(),
      sampler: sampler.createSchema(sampler.nai),
    }),
    Schema.object({
      sampler: sampler.createSchema(sampler.nai),
      model: Schema.union(models).loose().description('默认的生成模型。').default('nai'),
    }),
  ] as const),

  Schema.object({
    scale: Schema.computed(Schema.number(), options).description('默认对输入的服从度。').default(11),
    textSteps: Schema.computed(Schema.natural(), options).description('文本生图时默认的迭代步数。').default(28),
    imageSteps: Schema.computed(Schema.natural(), options).description('以图生图时默认的迭代步数。').default(50),
    maxSteps: Schema.computed(Schema.natural(), options).description('允许的最大迭代步数。').default(64),
    strength: Schema.computed(Schema.number(), options).min(0).max(1).description('默认的重绘强度。').default(0.7),
    resolution: Schema.computed(Schema.union([
      Schema.const('portrait').description('肖像 (768x512)'),
      Schema.const('landscape').description('风景 (512x768)'),
      Schema.const('square').description('方形 (640x640)'),
      Schema.object({
        width: Schema.natural().description('图片宽度。').default(640),
        height: Schema.natural().description('图片高度。').default(640),
      }).description('自定义'),
    ]), options).description('默认生成的图片尺寸。').default('portrait'),
    maxResolution: Schema.computed(Schema.natural(), options).description('允许生成的宽高最大值。').default(1024),
  }),

  PromptConfig,

  Schema.object({
    output: Schema.union([
      Schema.const('minimal').description('只发送图片'),
      Schema.const('default').description('发送图片和关键信息'),
      Schema.const('verbose').description('发送全部信息'),
    ]).description('输出方式。').default('default'),
    maxIterations: Schema.natural().description('允许的最大绘制次数。').default(1),
    maxRetryCount: Schema.natural().description('连接失败时最大的重试次数。').default(3),
    requestTimeout: Schema.number().role('time').description('当请求超过这个时间时会中止并提示超时。').default(Time.minute),
    recallTimeout: Schema.number().role('time').description('图片发送后自动撤回的时间 (设置为 0 以禁用此功能)。').default(0),
    maxConcurrency: Schema.number().description('单个频道下的最大并发数量 (设置为 0 以禁用此功能)。').default(0),
  }).description('高级设置'),
]) as Schema<Config>

interface Forbidden {
  pattern: string
  strict: boolean
}

export function parseForbidden(input: string) {
  return input.trim()
    .toLowerCase()
    .replace(/，/g, ',')
    .replace(/！/g, '!')
    .split(/(?:,\s*|\s*\n\s*)/g)
    .filter(Boolean)
    .map<Forbidden>((pattern: string) => {
      const strict = pattern.endsWith('!')
      if (strict) pattern = pattern.slice(0, -1)
      pattern = pattern.replace(/[^a-z0-9\u00ff-\uffff]+/g, ' ').trim()
      return { pattern, strict }
    })
}

const backslash = /@@__BACKSLASH__@@/g

export function parseInput(session: Session, input: string, config: Config, override: boolean, addDefault: boolean): string[] {
  if (!input) {
    return [
      null,
      [session.resolve(config.basePrompt), session.resolve(config.defaultPrompt)].join(','),
      session.resolve(config.negativePrompt)
    ]
  }

  input = input
    .replace(/\\\\/g, backslash.source)
    .replace(/，/g, ',')
    .replace(/（/g, '(')
    .replace(/）/g, ')')

  if (config.type === 'sd-webui') {
    input = input
      .split('\\{').map(s => s.replace(/\{/g, '(')).join('\\{')
      .split('\\}').map(s => s.replace(/\}/g, ')')).join('\\}')
  } else {
    input = input
      .split('\\(').map(s => s.replace(/\(/g, '{')).join('\\(')
      .split('\\)').map(s => s.replace(/\)/g, '}')).join('\\)')
  }

  input = input
    .replace(backslash, '\\')
    .replace(/_/g, ' ')

  if (session.resolve(config.latinOnly) && /[^\s\w"'“”‘’.,:|\\()\[\]{}<>-]/.test(input)) {
    return ['.latin-only']
  }

  const negative = []
  const placement = session.resolve(config.placement)
  const appendToList = (words: string[], input = '') => {
    const tags = input.split(/,\s*/g)
    if (placement === 'before') tags.reverse()
    for (let tag of tags) {
      tag = tag.trim().toLowerCase()
      if (!tag || words.includes(tag)) continue
      if (placement === 'before') {
        words.unshift(tag)
      } else {
        words.push(tag)
      }
    }
  }

  // extract negative prompts
  const capture = input.match(/(,\s*|\s+)(-u\s+|--undesired\s+|negative prompts?:\s*)([\s\S]+)/m)
  if (capture?.[3]) {
    input = input.slice(0, capture.index).trim()
    appendToList(negative, capture[3])
  }

  // remove forbidden words
  const forbidden = parseForbidden(session.resolve(config.forbidden))
  const positive = input.split(/,\s*/g).filter((word) => {
    // eslint-disable-next-line no-control-regex
    word = word.toLowerCase().replace(/[\x00-\x7f]/g, s => s.replace(/[^0-9a-zA-Z]/, ' ')).replace(/\s+/, ' ').trim()
    if (!word) return false
    for (const { pattern, strict } of forbidden) {
      if (strict && word.split(/\W+/g).includes(pattern)) {
        return false
      } else if (!strict && word.includes(pattern)) {
        return false
      }
    }
    return true
  }).map((word) => {
    if (/^<.+>$/.test(word)) return word.replace(/ /g, '_')
    return word.toLowerCase()
  })

  if (Math.max(getWordCount(positive), getWordCount(negative)) > (session.resolve(config.maxWords) || Infinity)) {
    return ['.too-many-words']
  }

  if (!override) {
    appendToList(positive, session.resolve(config.basePrompt))
    appendToList(negative, session.resolve(config.negativePrompt))
    if (addDefault) appendToList(positive, session.resolve(config.defaultPrompt))
  }

  return [null, positive.join(', '), negative.join(', ')]
}

function getWordCount(words: string[]) {
  return words.join(' ').replace(/[^a-z0-9]+/g, ' ').trim().split(' ').length
}
